<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Imports" data-toc-modified-id="Imports-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Imports</a></span></li><li><span><a href="#Get-Data" data-toc-modified-id="Get-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Get Data</a></span></li><li><span><a href="#Profile-the-Dataset" data-toc-modified-id="Profile-the-Dataset-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Profile the Dataset</a></span></li><li><span><a href="#Clean-the-Data" data-toc-modified-id="Clean-the-Data-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Clean the Data</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#-Fill-in-Missing-Neighborhood-Values" data-toc-modified-id="-Fill-in-Missing-Neighborhood-Values-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span> Fill in Missing Neighborhood Values</a></span></li><li><span><a href="#-Get-Year-and-Quarter" data-toc-modified-id="-Get-Year-and-Quarter-4.0.2"><span class="toc-item-num">4.0.2&nbsp;&nbsp;</span> Get Year and Quarter</a></span></li><li><span><a href="#Save" data-toc-modified-id="Save-4.0.3"><span class="toc-item-num">4.0.3&nbsp;&nbsp;</span>Save</a></span></li></ul></li></ul></li></ul></div>

<a id="section__top"></a>

<h2>Imports

In [51]:
import pandas as pd
import numpy as np
import pandas_profiling
import time
import datetime

import os
import sys
import re

module_path = os.path.abspath(os.path.join('./lib/'))

if module_path not in sys.path:   
    sys.path.append(module_path)
    
from utilities import *
from sodapy_dataset_reader import *

The sodapy_dataset_reader import contains a classes that use the
__[sodapy library ](https://pypi.org/project/sodapy/)__
to interface with the __[Socrata API ](https://dev.socrata.com)__
<br>
Many open data platforms are implemented using the Socrata (or SODA) API, including
__[SFData.gov](https://data.sfgov.org/)__, which is used for this project.

<h2>Get Data

In [2]:
#data source
source_domain = 'data.sfgov.org'
#dataset: eviction notices
dataset_id = '5cei-gny5'

#Instantiate Reader
spr = SodapyDatasetReader(source_domain, dataset_id)

#Get Meta Data
md = spr.get_metadata()

#get and print rowcount
rowcount = int(spr.get_row_count()[0]['count'])
print(f'Dataset contains {rowcount} rows')

#set limit to rowcount
df = spr.get_df(limit=rowcount)
df.shape

Dataset contains 40449 rows


(40449, 42)

In [3]:
#md

<h2>Profile the Dataset

The Pandas profiler is a utility function that expands on the initial data exploration 
steps one does when one first loads a new dataset. Code and documentation can be found here:
__[Pandas Profiling](https://github.com/pandas-profiling/pandas-profiling)__

In [4]:
profile = pandas_profiling.ProfileReport(df)

In [5]:
display(profile)

In [6]:
#output to file...
fn = f'profile-{dataset_id}.html'
profile.to_file(outputfile='../tmp/fn')

<h2>Clean the Data

[back to top](#section__top)

In [7]:
df.head()

:@computed_region_26cr_cadq :@computed_region_6ezc_tdp2  \
0                           8                         NaN   
1                          10                           1   
2                           3                         NaN   
3                           3                         NaN   
4                           3                         NaN   

  :@computed_region_6pnf_4xz7 :@computed_region_6qbp_sg9q  \
0                           1                          71   
1                           2                          20   
2                           2                          99   
3                           2                          99   
4                           2                          99   

  :@computed_region_9jxd_iqea :@computed_region_ajp5_b2md  \
0                         NaN                          41   
1                           6                          36   
2                         NaN                          23   
3                         NaN                          23   
4                         NaN                          23   

  :@computed_region_bh8s_q3mv :@computed_region_fyvs_ahh9  \
0                       28861                          40   
1                       28852                          36   
2                         308                          23   
3                         308                          23   
4                         308                          23   

  :@computed_region_h4ep_8xdi :@computed_region_p5aj_wyqh  ... non_payment  \
0                         NaN                           7  ...       False   
1                           1                          10  ...       False   
2                         NaN                           1  ...       False   
3                         NaN                           1  ...       False   
4                         NaN                           1  ...       False   

  nuisance other_cause  owner_move_in roommate_same_unit  state  \
0    False       False           True              False     CA   
1     True       False          False              False     CA   
2    False       False          False              False     CA   
3    False       False          False              False     CA   
4    False       False          False              False     CA   

   substantial_rehab supervisor_district unapproved_subtenant    zip  
0              False                   7                False  94112  
1              False                   6                False  94102  
2              False                   3                False  94133  
3              False                   3                False  94133  
4              False                   3                False  94133  

[5 rows x 42 columns]

In [8]:
df.shape

(40449, 42)

In [9]:
df.columns

Index([':@computed_region_26cr_cadq', ':@computed_region_6ezc_tdp2',
       ':@computed_region_6pnf_4xz7', ':@computed_region_6qbp_sg9q',
       ':@computed_region_9jxd_iqea', ':@computed_region_ajp5_b2md',
       ':@computed_region_bh8s_q3mv', ':@computed_region_fyvs_ahh9',
       ':@computed_region_h4ep_8xdi', ':@computed_region_p5aj_wyqh',
       ':@computed_region_qgnn_b9vv', ':@computed_region_rxqg_mtj9',
       ':@computed_region_yftq_j783', 'access_denial', 'address', 'breach',
       'capital_improvement', 'city', 'client_location', 'condo_conversion',
       'constraints_date', 'demolition', 'development', 'ellis_act_withdrawal',
       'eviction_id', 'failure_to_sign_renewal', 'file_date',
       'good_samaritan_ends', 'illegal_use', 'late_payments',
       'lead_remediation', 'neighborhood', 'non_payment', 'nuisance',
       'other_cause', 'owner_move_in', 'roommate_same_unit', 'state',
       'substantial_rehab', 'supervisor_district', 'unapproved_subtenant',
       'zip'],

drop computed columns per this link
https://support.socrata.com/hc/en-us/articles/360007155973-Handling-Computed-Columns-with-FME
Datasets that have location data that intersects a Spatial Lens Boundary will have computed columns. These columns are created and curated by the Socrata platform, not the user. They are used to perform the geographic join between the geocoded row and the underlying spatial lens polygon. The column names begin with the prefix :@computed_region_.

In [10]:
df = df.drop([':@computed_region_26cr_cadq', ':@computed_region_6ezc_tdp2',
       ':@computed_region_6pnf_4xz7', ':@computed_region_6qbp_sg9q',
       ':@computed_region_9jxd_iqea', ':@computed_region_ajp5_b2md',
       ':@computed_region_bh8s_q3mv', ':@computed_region_fyvs_ahh9',
       ':@computed_region_h4ep_8xdi', ':@computed_region_p5aj_wyqh',
       ':@computed_region_qgnn_b9vv', ':@computed_region_rxqg_mtj9',
       ':@computed_region_yftq_j783'], axis = 1)

df.shape

(40449, 29)

In [11]:
df.head()

access_denial                           address  breach  \
0          False        700 Block Of Faxon  Avenue   False   
1          False         100 Block Of Turk  Street   False   
2          False          200 Block Of Bay  Street   False   
3          False          200 Block Of Bay  Street   False   
4          False  200 Block Of North Point  Street   False   

   capital_improvement           city  \
0                False  San Francisco   
1                False  San Francisco   
2                False  San Francisco   
3                False  San Francisco   
4                False  San Francisco   

                                     client_location  condo_conversion  \
0  {'latitude': '37.726686880000024', 'longitude'...             False   
1  {'latitude': '37.78316342723367', 'longitude':...             False   
2  {'latitude': '37.805980597572834', 'longitude'...             False   
3  {'latitude': '37.805980597572834', 'longitude'...             False   
4  {'latitude': '37.80680623758749', 'longitude':...             False   

          constraints_date  demolition  development  ...  non_payment  \
0  2024-03-29T00:00:00.000       False        False  ...        False   
1                      NaN       False        False  ...        False   
2                      NaN       False        False  ...        False   
3                      NaN       False        False  ...        False   
4                      NaN       False        False  ...        False   

  nuisance  other_cause owner_move_in  roommate_same_unit  state  \
0    False        False          True               False     CA   
1     True        False         False               False     CA   
2    False        False         False               False     CA   
3    False        False         False               False     CA   
4    False        False         False               False     CA   

   substantial_rehab  supervisor_district unapproved_subtenant    zip  
0              False                    7                False  94112  
1              False                    6                False  94102  
2              False                    3                False  94133  
3              False                    3                False  94133  
4              False                    3                False  94133  

[5 rows x 29 columns]

In [12]:
df.isnull().sum()

access_denial                  0
address                        0
breach                         0
capital_improvement            0
city                           2
client_location             1400
condo_conversion               0
constraints_date           36428
demolition                     0
development                    0
ellis_act_withdrawal           0
eviction_id                    0
failure_to_sign_renewal        0
file_date                      0
good_samaritan_ends            0
illegal_use                    0
late_payments                  0
lead_remediation               0
neighborhood                1412
non_payment                    0
nuisance                       0
other_cause                    0
owner_move_in                  0
roommate_same_unit             0
state                          2
substantial_rehab              0
supervisor_district         1412
unapproved_subtenant           0
zip                          682
dtype: int64

Examine the neighborhoods. They can be matched with those in the main dataset. 

In [13]:
neighborhoods = set(df.neighborhood)

In [14]:
neighborhoods

{'Bayview Hunters Point',
 'Bernal Heights',
 'Castro/Upper Market',
 'Chinatown',
 'Excelsior',
 'Financial District/South Beach',
 'Glen Park',
 'Golden Gate Park',
 'Haight Ashbury',
 'Hayes Valley',
 'Inner Richmond',
 'Inner Sunset',
 'Japantown',
 'Lakeshore',
 'Lincoln Park',
 'Lone Mountain/USF',
 'Marina',
 'McLaren Park',
 'Mission',
 'Mission Bay',
 'Nob Hill',
 'Noe Valley',
 'North Beach',
 'Oceanview/Merced/Ingleside',
 'Outer Mission',
 'Outer Richmond',
 'Pacific Heights',
 'Portola',
 'Potrero Hill',
 'Presidio',
 'Presidio Heights',
 'Russian Hill',
 'Seacliff',
 'South of Market',
 'Sunset/Parkside',
 'Tenderloin',
 'Treasure Island',
 'Twin Peaks',
 'Visitacion Valley',
 'West of Twin Peaks',
 'Western Addition',
 nan}

Fill missing values for columns we are likely to use in the analysis. Do not drop any for now.

In [23]:
fill_cols = {'neighborhood':'.', 
             'supervisor_district':'.',
             'zip':'.'
            }

In [25]:
#df.fillna(fill_cols).isnull().sum()
df.fillna(fill_cols, inplace=True)

In [26]:
df.isnull().sum()

access_denial                  0
address                        0
breach                         0
capital_improvement            0
city                           2
client_location             1400
condo_conversion               0
constraints_date           36428
demolition                     0
development                    0
ellis_act_withdrawal           0
eviction_id                    0
failure_to_sign_renewal        0
file_date                      0
good_samaritan_ends            0
illegal_use                    0
late_payments                  0
lead_remediation               0
neighborhood                   0
non_payment                    0
nuisance                       0
other_cause                    0
owner_move_in                  0
roommate_same_unit             0
state                          2
substantial_rehab              0
supervisor_district            0
unapproved_subtenant           0
zip                            0
dtype: int64

Examine the city names. There are errors, this column won't be used in the analysis. This just confirms the data applies only to the city of San Francisco.

In [27]:
df.city.value_counts()

San Francisco      40430
San Franicsco          4
San Franisco           2
Sn Francisco           2
San Francisc           1
158an Francisco        1
San ‘Francisco         1
San Frncisco           1
San Francicso          1
San Francisoc          1
San Frnacisco          1
3/9/2017               1
San Franciso           1
Name: city, dtype: int64

In [20]:
df.columns

Index(['access_denial', 'address', 'breach', 'capital_improvement', 'city',
       'client_location', 'condo_conversion', 'constraints_date', 'demolition',
       'development', 'ellis_act_withdrawal', 'eviction_id',
       'failure_to_sign_renewal', 'file_date', 'good_samaritan_ends',
       'illegal_use', 'late_payments', 'lead_remediation', 'neighborhood',
       'non_payment', 'nuisance', 'other_cause', 'owner_move_in',
       'roommate_same_unit', 'state', 'substantial_rehab',
       'supervisor_district', 'unapproved_subtenant', 'zip'],
      dtype='object')

Confirm that the client location field has the coordinate information.  We need to extract that to lat and lon.

In [28]:
df.iloc[0].client_location

{'latitude': '37.726686880000024',
 'longitude': '-122.4598938584836',
 'human_address': '{"address": "", "city": "", "state": "", "zip": ""}'}

The following code 

In [29]:
from LatLong import *
def convert_location(id, nm, loc):
    try:
        lat = float(loc['latitude'])
        lon = float(loc['longitude'])
    except:
        #print(f'an error occurred extracting coordinates on {id}, {nm}')
        lat, lon = 0,0
    #print(f'returning lat: {lat}, lon: {lon}')
    return lat, lon

In [30]:
start_execution = time.time()

df[['lat', 'lon']] = df.apply(lambda row: convert_location( row.eviction_id, 
                                     row.eviction_id,
                                     row.client_location), 
                                       axis=1, result_type='expand')
end_execution = time.time()
print(f'Elapsed time: {end_execution - start_execution}')


Elapsed time: 10.2966148853302


Use .describe() to verify the lat and lon columns

In [35]:
df.lon.describe(), df.lat.describe()

(count    40449.000000
 mean      -118.199439
 std         22.381036
 min       -122.511077
 25%       -122.452921
 50%       -122.428534
 75%       -122.413833
 max          0.000000
 Name: lon, dtype: float64, count    40449.000000
 mean        36.456004
 std          6.902975
 min          0.000000
 25%         37.743222
 50%         37.764445
 75%         37.781983
 max         37.830455
 Name: lat, dtype: float64)

In [ ]:
# ts = get_timestamp()
# df.to_csv(f'../tmp/evic_cleanstge1_{ts}.csv', index=False)

In [ ]:
# df = pd.read_csv('../tmp/evic_cleanstge1_2019201951136955025.csv', low_memory = False)

<h4> Fill in Missing Neighborhood Values

All attempts to fill in missing neighborhoods failed. May be a sign of bad coordinates. These rows will not be included in the analysis.

In [37]:
from LatLong import *

def get_neighborhood(lat,lon):
    try:
        ll = LatLong(lat,lon)
        neighborhood = ll.get_neighborhood()
    except:
        return '.'
    else:
        return neighborhood

In [38]:
#This is a sample of some the attempts
#ret = get_neighborhood(37.806, -122.411)
#37.70796257157375 -122.4635364761915
#37.707921903319395 -122.4287166819376
ret = get_neighborhood(37.70792, -122.4287)
#ret = get_neighborhood(37.806, -122.411)
#ret = get_neighborhood(df.iloc[0]['lat'],df.iloc[0]['lon'])
ret

'.'

In [40]:
start_execution = time.time()
df['new_neighborhood'] = df.apply(lambda row: get_neighborhood(row.lat, row.lon) if  (row.neighborhood == '.') 
                                                                 &   (row.lat > 30)                                                 
                                  else row.neighborhood,
                                  axis=1)
end_execution = time.time()
print(f'Elapsed time: {end_execution - start_execution}')

Elapsed time: 7.864480018615723


In [41]:
df.neighborhood.value_counts()

Mission                           4383
Tenderloin                        2545
Sunset/Parkside                   2538
Outer Richmond                    2039
Castro/Upper Market               1814
Lakeshore                         1621
Hayes Valley                      1472
South of Market                   1420
.                                 1412
Nob Hill                          1378
Haight Ashbury                    1293
Marina                            1250
Excelsior                         1210
Noe Valley                        1194
Inner Sunset                      1174
Bernal Heights                    1172
Pacific Heights                   1134
Inner Richmond                    1105
Bayview Hunters Point             1100
Russian Hill                       967
North Beach                        949
Oceanview/Merced/Ingleside         837
Lone Mountain/USF                  825
West of Twin Peaks                 741
Outer Mission                      682
Western Addition         

Note that there are 1,412 eviction notice records for which a specific neighborhood cannot be found.

<h4> Get Year and Quarter

Extract year and quarter into separate columns to be used when combining with business data.  Then display the data to ensure it was done correctly.

In [42]:
# #Convert date into quarter string
def get_year_and_quarter(dt):
    dt = pd.Timestamp(dt)
    year = dt.year        
    quarter = dt.quarter
    #print(f'{dt}, {year}, {quarter}, {dt.year}-Q{dt.quarter}' )
    return year, quarter, f'{dt.year}-Q{dt.quarter}'

In [43]:
df[['year', 'quarter', 'yq']] = df.apply(lambda row: get_year_and_quarter(row.file_date),
                       axis=1,  result_type='expand')
                       

In [44]:
df.head()

access_denial                           address  breach  \
0          False        700 Block Of Faxon  Avenue   False   
1          False         100 Block Of Turk  Street   False   
2          False          200 Block Of Bay  Street   False   
3          False          200 Block Of Bay  Street   False   
4          False  200 Block Of North Point  Street   False   

   capital_improvement           city  \
0                False  San Francisco   
1                False  San Francisco   
2                False  San Francisco   
3                False  San Francisco   
4                False  San Francisco   

                                     client_location  condo_conversion  \
0  {'latitude': '37.726686880000024', 'longitude'...             False   
1  {'latitude': '37.78316342723367', 'longitude':...             False   
2  {'latitude': '37.805980597572834', 'longitude'...             False   
3  {'latitude': '37.805980597572834', 'longitude'...             False   
4  {'latitude': '37.80680623758749', 'longitude':...             False   

          constraints_date  demolition  development  ...  substantial_rehab  \
0  2024-03-29T00:00:00.000       False        False  ...              False   
1                      NaN       False        False  ...              False   
2                      NaN       False        False  ...              False   
3                      NaN       False        False  ...              False   
4                      NaN       False        False  ...              False   

  supervisor_district  unapproved_subtenant    zip        lat         lon  \
0                   7                 False  94112  37.726687 -122.459894   
1                   6                 False  94102  37.783163 -122.411599   
2                   3                 False  94133  37.805981 -122.411141   
3                   3                 False  94133  37.805981 -122.411141   
4                   3                 False  94133  37.806806 -122.411309   

     new_neighborhood  year quarter       yq  
0  West of Twin Peaks  2019       1  2019-Q1  
1          Tenderloin  2019       1  2019-Q1  
2         North Beach  2019       1  2019-Q1  
3         North Beach  2019       1  2019-Q1  
4         North Beach  2019       1  2019-Q1  

[5 rows x 35 columns]

In [45]:
df[['file_date', 'year', 'quarter', 'yq']].tail(15)

file_date  year  quarter       yq
40434  1997-01-03T00:00:00.000  1997        1  1997-Q1
40435  1997-01-03T00:00:00.000  1997        1  1997-Q1
40436  1997-01-03T00:00:00.000  1997        1  1997-Q1
40437  1997-01-03T00:00:00.000  1997        1  1997-Q1
40438  1997-01-03T00:00:00.000  1997        1  1997-Q1
40439  1997-01-02T00:00:00.000  1997        1  1997-Q1
40440  1997-01-02T00:00:00.000  1997        1  1997-Q1
40441  1997-01-02T00:00:00.000  1997        1  1997-Q1
40442  1997-01-02T00:00:00.000  1997        1  1997-Q1
40443  1997-01-02T00:00:00.000  1997        1  1997-Q1
40444  1997-01-02T00:00:00.000  1997        1  1997-Q1
40445  1997-01-02T00:00:00.000  1997        1  1997-Q1
40446  1997-01-02T00:00:00.000  1997        1  1997-Q1
40447  1997-01-02T00:00:00.000  1997        1  1997-Q1
40448  1997-01-02T00:00:00.000  1997        1  1997-Q1

In [46]:
df.year.describe()

count    40449.000000
mean      2006.821256
std          6.860986
min       1997.000000
25%       2000.000000
50%       2006.000000
75%       2013.000000
max       2019.000000
Name: year, dtype: float64

In [47]:
df.quarter.describe()

count    40449.000000
mean         2.473188
std          1.095643
min          1.000000
25%          2.000000
50%          2.000000
75%          3.000000
max          4.000000
Name: quarter, dtype: float64

<h4>Save

Save the data in the tmp directory. It will be combined with other data to create the final dataset for analysis.

In [49]:
ts = get_timestamp()
df.to_csv(f'../tmp/evic_cleanstge2_{ts}.csv', index=False)